### PR TITLE
Introduce custom status item view on OS X.

### LINKED ISSUE #1499: [OS X] Tray `right-clicked` event
Many tray (menubar) apps on OS X have different behavior on click and right-click.

-

Some don't show a menu when clicked, but instead changes the icon and does an action, but shows a menu for eg. Preferences and Quit on right-click:

Normal:

screen shot 2013-07-29 at 21 27 16

Clicked:

screen shot 2013-07-29 at 21 27 21

Right-clicked:

screen shot 2013-07-29 at 21 27 38

-

Unfortunately there isn't a built-in method to do right-click directly on a NSStatusItem, but it's easy to achieve it with a custom view. There's a handy library that handles it for you: https://github.com/heardrwt/RHStatusItemView

-

It should probably also trigger the right click event on Option+click.

-

*This might be possible on Linux/Windows too, I haven't researched, but I'm interested in OS X support.*

### PR DESCRIPTION
**Pull request by hokein**
This PR introduces a custom status item view in NSStatusItem, so that we can easily implement some features in tray:

* Implement `right-clicked` event on OS X and Windows. As a side effect, the 'clicked' event will also be emitted even the tray has context menu.
* Implement `popContextMenu` API on OS X and Windows. 
* Implement `dragged-files` event to support drag-and-drop feature on OS X.

Fixes #1499,  #1393, #1975, #1549.

### CODE REVIEW

File: atom/browser/ui/tray_icon_cocoa.mm [hunk 1]
```diff
@@ -9,81 +9,216 @@
 #include "ui/gfx/image/image.h"
 #include "ui/gfx/screen.h"
 
-@interface StatusItemController : NSObject {
+const CGFloat kStatusItemLength = 26;
```
Comment (zcbenz) [inline atom/browser/ui/tray_icon_cocoa.mm]: Static variables should be put int anonymous namespace.

File: atom/browser/ui/tray_icon_cocoa.mm [hunk 2]
```diff
@@ -9,81 +9,216 @@
...
+                       fraction:1];
+  } else {
+    [image_ drawInRect:NSInsetRect(iconRect, kMargin, kMargin)
+              fromRect:NSZeroRect
+             operation:NSCompositeSourceOver
+              fraction:1];
+  }
+}
+
+- (CGFloat) titleWidth {
```
Comment (zcbenz) [inline atom/browser/ui/tray_icon_cocoa.mm]: `(CGFloat) titleWidth` => `(CGFloat)titleWidth`

File: atom/browser/ui/tray_icon_cocoa.mm [hunk 3]
```diff
@@ -9,81 +9,216 @@
...
+  if (!title_) return 0;
+  NSAttributedString* attributes =
+      [[NSAttributedString alloc] initWithString:title_
+                                      attributes:[self titleAttributes]];
+  return [attributes size].width;
+}
+
+- (NSDictionary *)titleAttributes {
```
Comment (zcbenz) [inline atom/browser/ui/tray_icon_cocoa.mm]: `NSDictionary *` to `NSDictionary*`

File: atom/browser/ui/tray_icon_cocoa.mm [hunk 4]
```diff
@@ -9,81 +9,216 @@
...
+  title_.reset([title copy]);
 }
 
-- (void)handleDoubleClick:(id)sender {
-  trayIcon_->NotifyDoubleClicked();
+- (void)setMenuController:(AtomMenuController*)menu {
+  menuController_ = menu;
 }
 
+-(void)mouseDown:(NSEvent *)event {
```
Comment (zcbenz) [inline atom/browser/ui/tray_icon_cocoa.mm]: `-(void)mouseDown:(NSEvent *)event` => `- (void)mouseDown:(NSEvent*)event`

File: atom/browser/ui/tray_icon_cocoa.mm [hunk 5]
```diff
@@ -9,81 +9,216 @@
...
+  inMouseEventSequence_ = YES;
+  [self setNeedsDisplay:YES];
+}
+
+-(void)mouseUp:(NSEvent *)event {
```
Comment (zcbenz) [inline atom/browser/ui/tray_icon_cocoa.mm]: `-(void)mouseUp:(NSEvent *)event` => `- (void)mouseUp:(NSEvent*)event`.

File: atom/browser/ui/tray_icon_cocoa.mm [hunk 6]
```diff
@@ -9,81 +9,216 @@
...
+    // When it returns, we need to redraw the tray icon to not show highlight.
+    [self setNeedsDisplay:YES];
+  }
+}
+
+- (void)rightMouseUp:(NSEvent*)event {
+  trayIcon_->NotifyRightClicked([self getBoundsFromEvent:event]);
+}
+
+-(BOOL) shouldHighlight {
```
Comment (zcbenz) [inline atom/browser/ui/tray_icon_cocoa.mm]: `-(BOOL) shouldHighlight` => `- (BOOL)shouldHighlight`

File: atom/browser/ui/tray_icon_cocoa.mm [hunk 7]
```diff
@@ -9,81 +9,216 @@
...
+  BOOL is_menu_open = [menuController_ isMenuOpen];
+  return isHighlightEnable_ && (inMouseEventSequence_ || is_menu_open);
+}
+
+-(gfx::Rect) getBoundsFromEvent:(NSEvent*)event {
```
Comment (zcbenz) [inline atom/browser/ui/tray_icon_cocoa.mm]: `-(gfx::Rect) getBoundsFromEvent` => `- (gfx::Rect)getBoundsFromEvent`.

File: atom/browser/ui/tray_icon_cocoa.mm [hunk 8]
```diff
@@ -9,81 +9,216 @@
...
+  ///  ----------------
+  BOOL highlight = [self shouldHighlight];
+  CGFloat titleWidth = [self titleWidth];
+  // Calculate the total icon bounds.
+  NSRect statusItemBounds = NSMakeRect(0,
+                                       0,
+                                       kStatusItemLength + titleWidth,
+                                       [[statusItem_ statusBar] thickness]);
+  [statusItem_ drawStatusBarBackgroundInRect:statusItemBounds
+                               withHighlight:highlight];
```
Comment (zcbenz) [inline atom/browser/ui/tray_icon_cocoa.mm]: I didn't check the official behavior, but I think we shouldn't draw draw background when there is alternative image?

**Comment by zcbenz** [comment]
I haven't tested but I think we should redraw the status item after the `setImage`/`setTitle` calls?

**Comment by zcbenz** [comment]
And does the status item show correctly with dark menu bar?

File: atom/browser/ui/tray_icon_cocoa.mm (hunk 1 above)
Comment (hokein) [inline atom/browser/ui/tray_icon_cocoa.mm]: Done.

File: atom/browser/ui/tray_icon_cocoa.mm (hunk 2 above)
Comment (hokein) [inline atom/browser/ui/tray_icon_cocoa.mm]: Done.

File: atom/browser/ui/tray_icon_cocoa.mm (hunk 3 above)
Comment (hokein) [inline atom/browser/ui/tray_icon_cocoa.mm]: Done.

File: atom/browser/ui/tray_icon_cocoa.mm (hunk 4 above)
Comment (hokein) [inline atom/browser/ui/tray_icon_cocoa.mm]: Done.

File: atom/browser/ui/tray_icon_cocoa.mm (hunk 5 above)
Comment (hokein) [inline atom/browser/ui/tray_icon_cocoa.mm]: Done.

File: atom/browser/ui/tray_icon_cocoa.mm (hunk 6 above)
Comment (hokein) [inline atom/browser/ui/tray_icon_cocoa.mm]: Done.

File: atom/browser/ui/tray_icon_cocoa.mm (hunk 7 above)
Comment (hokein) [inline atom/browser/ui/tray_icon_cocoa.mm]: Done.

File: atom/browser/ui/tray_icon_cocoa.mm (hunk 8 above)
Comment (hokein) [inline atom/browser/ui/tray_icon_cocoa.mm]: Oops, I miss it. Done.

Comment (hokein) [inline atom/browser/ui/tray_icon_cocoa.mm]: This behavior is consitent with the official. The official behavior will also highlight when alternative image is set and hightlight is enable.

**Comment by hokein** [comment]
> I haven't tested but I think we should redraw the status item after the setImage/setTitle calls?

Oops. I have missed them. Done now.

> And does the status item show correctly with dark menu bar?

Haven't test it yet due to my OS X version is 10.9.

File: atom/browser/api/atom_api_tray.cc [hunk 9]
```diff
@@ -60,6 +60,14 @@ void Tray::OnBalloonClosed() {
   Emit("balloon-closed");
 }
 
+void Tray::OnRightClicked(const gfx::Rect& bounds) {
+  Emit("right-clicked", bounds);
+}
+
+void Tray::OnDraggedFiles(const std::vector<std::string>& files) {
+  Emit("dragged-files", files);
```
Comment (zcbenz) [inline atom/browser/api/atom_api_tray.cc]: `drop-files` might be a better name to match DOM's `drop` event.

Comment (hokein) [inline atom/browser/api/atom_api_tray.cc]: Done. Also having test in dark menu bar, the status item shows correctly.

File: atom/browser/api/atom_api_tray.cc [hunk 10]
```diff
@@ -60,6 +60,14 @@ void Tray::OnBalloonClosed() {
   Emit("balloon-closed");
 }
 
+void Tray::OnRightClicked(const gfx::Rect& bounds) {
+  Emit("right-clicked", bounds);
+}
+
+void Tray::OnDragFiles(const std::vector<std::string>& files) {
+  Emit("drag-files", files);
```
Comment (zcbenz) [inline atom/browser/api/atom_api_tray.cc]: As stated by DOM's docs:

> The drag event is fired when an element or text selection is being dragged (every few hundred milliseconds)
> The drop event is fired when an element or text selection is dropped on a valid drop target.

`drop` should be better than `drag` in this case?

Comment (hokein) [inline atom/browser/api/atom_api_tray.cc]: :+1:  Done.

**Comment by zcbenz** [comment]
:+1:

**Comment by jorangreef** [comment]
Thanks

**Comment by fourpixels** [comment]
Hey guys,
Can this be implemented for Windows too?

**Comment by kemuri** [comment]
Is it possible to drag&drop (URL) for example? As far as I see, right now only files are allowed :(

**Comment by black-snow** [comment]
Seems like it's not optional to pass a Menu to `Tray.popUpContextMenu()` (on windows at least).

**Pull request merged**